Core/src/Detector/CylindricalDetectorHelper.cpp

```diff
@@ -0,0 +1,1115 @@
+
+  if (refValues[0u] > 0.) {
+    // Inner cylinder
+    pReplacements.push_back(createCylinderReplacement(
+        refTransform, refValues[0u], {-refValues[2u], refValues[2u]},
+        phiBoundaries, 3u, Acts::NavigationDirection::Forward));
+  }
+
+  // Attach the new volume multi links
+  attachDetectorVolumeUpdators(gctx, volumes, pReplacements);
+  // Exchange the portals of the volumes
+  ACTS_VERBOSE("Portals of " << volumes.size() << " volumes need updating.");
+  for (auto& iv : volumes) {
+    ACTS_VERBOSE("- update portals of volume '" << iv->name() << "'.");
+    for (auto& [p, i, dir, boundaries, binning] : pReplacements) {
+      iv->updatePortal(p, static_cast<unsigned int>(i));
+      // Fill the map
+      protoContainer[i] = p;
+    }
+  }
+
+  // Done.
+  return protoContainer;
+}
+
+Acts::Experimental::ProtoContainer Acts::Experimental::wrapDetectorVolumesInZR(
+    const Acts::GeometryContext& gctx,
+    std::vector<std::shared_ptr<Acts::Experimental::DetectorVolume>>& volumes,
+    Acts::Logging::Level logLevel) {
+  // The local logger
+  ACTS_LOCAL_LOGGER(getDefaultLogger("CylindricalDetectorHelper", logLevel));
+
+  ACTS_DEBUG("Wrapping volumes in Z-R.");
+
+  // Minimal set of checks
+  if (volumes.size() != 2u) {
+    throw std::invalid_argument(
+        "Wrapping the detector volume requires exactly 2 volumes.");
+  }
+
+  // Return the new container
+  ProtoContainer protoContainer;
+
+  // Keep the outer shells
+  protoContainer[0u] = volumes[1u]->portalPtrs()[0u];
+  protoContainer[1u] = volumes[1u]->portalPtrs()[1u];
+  protoContainer[2u] = volumes[1u]->portalPtrs()[2u];
+
+  // Fuse outer cover of first with inner cylinder of wrapping volume
+  auto& keepCover = volumes[0u]->portalPtrs()[2u];
+  auto& wasteCover = volumes[1u]->portalPtrs()[3u];
+  keepCover->fuse(wasteCover);
+  volumes[1u]->updatePortal(keepCover, 3u);
+
+  // Stitch sides - negative
+  auto& keepDiscN = volumes[1u]->portalPtrs()[4u];
+  auto& wasteDiscN = volumes[0u]->portalPtrs()[0u];
+  keepDiscN->fuse(wasteDiscN);
+  volumes[0u]->updatePortal(keepDiscN, 0u);
+
+  // Stich sides - positive
+  auto& keepDiscP = volumes[0u]->portalPtrs()[1u];
+  auto& wasteDiscP = volumes[1u]->portalPtrs()[5u];
+  keepDiscP->fuse(wasteDiscP);
+  volumes[1u]->updatePortal(keepDiscP, 5u);
+
+  // If needed, insert new cylinder
+  if (volumes[0u]->portalPtrs().size() == 4u and
+      volumes[1u]->portalPtrs().size() == 8u) {
+    // We need a new cylinder spanning over the entire inner tube
+    ActsScalar hlZ =
+        volumes[0u]
+            ->volumeBounds()
+            .values()[Acts::CylinderVolumeBounds::BoundValues::eHalfLengthZ];
+    ActsScalar HlZ =
+        volumes[1u]->volumeBounds().values()
+            [Acts::CutoutCylinderVolumeBounds::BoundValues::eHalfLengthZ];
+    ActsScalar innerR =
+        volumes[0u]
+            ->volumeBounds()
+            .values()[Acts::CylinderVolumeBounds::BoundValues::eMinR];
```

Comment:
Do we assume innerR for the inner volume is the same as for the outer volume? If not, we need to build 3 portals, not only one, don't we?

Reply:
Yes, we assume everything is synchronised at this step, that's the job of the `ProtoDetector`.